File: LeetCode/Easy/0232_Implement_Queue_Using_Stacks/Solution.cpp
// 2022-12-15
class MyQueue {
public:
/*
    implement a queue using 2 stacks

    notes:
    - amortized O(1) operations
    - using 2 stacks
*/
    deque<int> s1;
    deque<int> s2;
    MyQueue() {
    }
    
    void push(int x) {
        s1.push_back(x);
    }
    
    int pop() {
        int val = peek();
        s2.pop_back();
        return val;
    }
    
    int peek() {
        if (s2.empty()) {
            while (!s1.empty()) {
                s2.push_back(s1.back());
                s1.pop_back();
            }
        }
        return s2.back();
    }
    
    bool empty() {
        return s1.empty() && s2.empty();
    }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */

File: LeetCode/Easy/0232_Implement_Queue_Using_Stacks/Solution.java
// 2021-11-16
class MyQueue {
    
    Deque<Integer> queue;
    Deque<Integer> stack;
    
    public MyQueue() {
        queue = new ArrayDeque<Integer>();
        stack = new ArrayDeque<Integer>();
    }
    
    public void push(int x) {
        stack.addFirst(x);
    }
    
    public int pop() {
        shiftStacks(stack, queue);
        int output = queue.pop();
        shiftStacks(queue, stack);
        return output;
    }
    
    private void shiftStacks(Deque<Integer> stack1, Deque<Integer> stack2) {
        while (!stack1.isEmpty()) {
            stack2.addFirst(stack1.pop());
        }
    }
    
    public int peek() {
        shiftStacks(stack, queue);
        int output = queue.getFirst();
        shiftStacks(queue, stack);
        return output;
    }
    
    public boolean empty() {
        return stack.size() == 0;
    }
    
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */